client/allocrunner/taskrunner/artifact_hook_test.go

```diff
@@ -158,3 +159,230 @@ func TestTaskRunner_ArtifactHook_PartialDone(t *testing.T) {
 }
+
+// TestTaskRunner_ArtifactHook_ConcurrentDownload asserts that the artifact hook
+// download multiple files concurrently. this is a successful test without any errors.
+func TestTaskRunner_ArtifactHook_ConcurrentDownloadSuccess(t *testing.T) {
+	t.Parallel()
+
+	me := &mockEmitter{}
+	artifactHook := newArtifactHook(me, testlog.HCLogger(t))
+
+	// Create a source directory all 7 artifacts
+	srcdir, err := ioutil.TempDir("", "nomadtest-src")
+	require.NoError(t, err)
+	defer func() {
+		require.NoError(t, os.RemoveAll(srcdir))
+	}()
+
+	numOfFiles := 7
+	for i := 0; i < numOfFiles; i++ {
+		file := filepath.Join(srcdir, fmt.Sprintf("file%d.txt", i))
+		require.NoError(t, ioutil.WriteFile(file, []byte{byte(i)}, 0644))
+	}
+
+	// Test server to serve the artifacts
+	ts := httptest.NewServer(http.FileServer(http.Dir(srcdir)))
+	defer ts.Close()
+
+	// Create the target directory.
+	destdir, err := ioutil.TempDir("", "nomadtest-dest")
+	require.NoError(t, err)
+	defer func() {
+		require.NoError(t, os.RemoveAll(destdir))
+	}()
+
+	req := &interfaces.TaskPrestartRequest{
+		TaskEnv: taskenv.NewTaskEnv(nil, nil, nil, nil, destdir, ""),
+		TaskDir: &allocdir.TaskDir{Dir: destdir},
+		Task: &structs.Task{
+			Artifacts: []*structs.TaskArtifact{
+				{
+					GetterSource: ts.URL + "/file0.txt",
+					GetterMode:   structs.GetterModeAny,
+				},
+				{
+					GetterSource: ts.URL + "/file1.txt",
+					GetterMode:   structs.GetterModeAny,
+				},
+				{
+					GetterSource: ts.URL + "/file2.txt",
+					GetterMode:   structs.GetterModeAny,
+				},
+				{
+					GetterSource: ts.URL + "/file3.txt",
+					GetterMode:   structs.GetterModeAny,
+				},
+				{
+					GetterSource: ts.URL + "/file4.txt",
+					GetterMode:   structs.GetterModeAny,
+				},
+				{
+					GetterSource: ts.URL + "/file5.txt",
+					GetterMode:   structs.GetterModeAny,
+				},
+				{
+					GetterSource: ts.URL + "/file6.txt",
+					GetterMode:   structs.GetterModeAny,
+				},
+			},
+		},
+	}
+
+	resp := interfaces.TaskPrestartResponse{}
+
+	// start the hook
+	err = artifactHook.Prestart(context.Background(), req, &resp)
+
+	require.NoError(t, err)
+	require.True(t, resp.Done)
+	require.Len(t, resp.State, 7)
+	require.Len(t, me.events, 1)
+	require.Equal(t, structs.TaskDownloadingArtifacts, me.events[0].Type)
+
+	// Assert all files downloaded properly
+	files, err := filepath.Glob(filepath.Join(destdir, "*.txt"))
+	require.NoError(t, err)
+	require.Len(t, files, 7)
+	sort.Strings(files)
+	require.Contains(t, files[0], "file0.txt")
+	require.Contains(t, files[1], "file1.txt")
+	require.Contains(t, files[2], "file2.txt")
+	require.Contains(t, files[3], "file3.txt")
+	require.Contains(t, files[4], "file4.txt")
+	require.Contains(t, files[5], "file5.txt")
+	require.Contains(t, files[6], "file6.txt")
+
+	// Stop the test server entirely and assert that re-running works
+	ts.Close()
+}
+
+// TestTaskRunner_ArtifactHook_ConcurrentDownload asserts that the artifact hook
```

Comment:
Wrong test name in comment: TestTaskRunner_ArtifactHook_ConcurrentDownloadFailure

Reply:
resolved via 1ff65dc2f